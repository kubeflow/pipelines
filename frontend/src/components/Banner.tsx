/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import ErrorIcon from '@material-ui/icons/Error';
import WarningIcon from '@material-ui/icons/Warning';
import { classes, stylesheet } from 'typestyle';

import { color, commonCss, spacing } from '../Css';

export type Mode = 'error' | 'warning';

export const css = stylesheet({
  banner: {
    border: `1px solid ${color.divider}`,
    boxSizing: 'border-box',
    justifyContent: 'space-between',
    margin: spacing.units(-1),
    minHeight: '50px',
    padding: spacing.units(-4),
  },
  button: {
    color: color.secondaryText,
    maxHeight: '32px',
    minWidth: '75px',
  },
  detailsButton: {
    backgroundColor: color.background,
  },
  icon: {
    height: '18px',
    padding: spacing.units(-4),
    width: '18px',
  },
  message: {
    alignItems: 'center',
    display: 'flex',
  },
  refreshButton: {
    backgroundColor: color.background,
  },
  troubleShootingLink: {
    alignItems: 'center',
    color: color.theme,
    fontWeight: 'bold',
    padding: spacing.units(-4),
  },
});

export interface BannerProps {
  additionalInfo?: string;
  message?: string;
  mode?: Mode;
  showTroubleshootingGuideLink?: boolean;
  refresh?: () => void;
}

interface BannerState {
  dialogOpen: boolean;
}

class Banner extends React.Component<BannerProps, BannerState> {
  constructor(props: any) {
    super(props);

    this.state = {
      dialogOpen: false,
    };
  }

  public render(): JSX.Element {
    // Default to error styles.
    let bannerModeCss = stylesheet({
      mode: { backgroundColor: color.errorBg, color: color.errorText },
    });
    let bannerIcon = <ErrorIcon className={css.icon} />;
    let dialogTitle = 'An error occurred';

    switch (this.props.mode) {
      case 'error':
        bannerModeCss = stylesheet({
          mode: { backgroundColor: color.errorBg, color: color.errorText },
        });
        bannerIcon = <ErrorIcon className={css.icon} />;
        dialogTitle = 'An error occurred';
        break;
      case 'warning':
        bannerModeCss = stylesheet({
          mode: { backgroundColor: color.warningBg, color: color.warningText },
        });
        bannerIcon = <WarningIcon className={css.icon} />;
        dialogTitle = 'Warning';
        break;
      default:
        // Already set defaults above.
        break;
    }

    return (
      <div className={classes(commonCss.flex, css.banner, bannerModeCss.mode)}>
        <div className={css.message}>
          {bannerIcon}
          {this.props.message}
        </div>
        <div className={commonCss.flex}>
          {this.props.showTroubleshootingGuideLink && (
            <a
              className={css.troubleShootingLink}
              href='https://www.kubeflow.org/docs/pipelines/troubleshooting'
            >
              Troubleshooting guide
            </a>
          )}
          {this.props.additionalInfo && (
            <Button
              className={classes(css.button, css.detailsButton)}
              onClick={this._showAdditionalInfo.bind(this)}
            >
              Details
            </Button>
          )}
          {this.props.refresh && (
            <Button
              className={classes(css.button, css.refreshButton)}
              onClick={this._refresh.bind(this)}
            >
              Refresh
            </Button>
          )}
        </div>

        {this.props.additionalInfo && (
          <Dialog open={this.state.dialogOpen} onClose={this._dialogClosed.bind(this)}>
            <DialogTitle>{dialogTitle}</DialogTitle>
            <DialogContent className={commonCss.prewrap}>{this.props.additionalInfo}</DialogContent>
            <DialogActions>
              <Button id='dismissDialogBtn' onClick={this._dialogClosed.bind(this)}>
                Dismiss
              </Button>
            </DialogActions>
          </Dialog>
        )}
      </div>
    );
  }

  private _dialogClosed(): void {
    this.setState({ dialogOpen: false });
  }

  private _showAdditionalInfo(): void {
    this.setState({ dialogOpen: true });
  }

  private _refresh(): void {
    this.props.refresh!();
  }
}

export default Banner;
