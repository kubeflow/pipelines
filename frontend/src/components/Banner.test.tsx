/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';

import { shallow } from 'enzyme';
import Banner, { css } from './Banner';

describe('Banner', () => {
  it('defaults to error mode', () => {
    const tree = shallow(<Banner message={'Some message'} />);
    expect(tree).toMatchSnapshot();
  });

  it('uses error mode when instructed', () => {
    const tree = shallow(<Banner message={'Some message'} mode={'error'} />);
    expect(tree).toMatchSnapshot();
  });

  it('uses warning mode when instructed', () => {
    const tree = shallow(<Banner message={'Some message'} mode={'warning'} />);
    expect(tree).toMatchSnapshot();
  });

  it('shows "Details" button and has dialog when there is additional info', () => {
    const tree = shallow(<Banner message={'Some message'} additionalInfo={'More info'} />);
    expect(tree).toMatchSnapshot();
  });

  it('shows "Refresh" button when passed a refresh function', () => {
    const tree = shallow(
      <Banner
        message={'Some message'}
        refresh={() => {
          /* do nothing */
        }}
      />,
    );
    expect(tree).toMatchSnapshot();
  });

  it('shows troubleshooting link instructed by prop', () => {
    const tree = shallow(<Banner message='Some message' showTroubleshootingGuideLink={true} />);
    expect(tree).toMatchInlineSnapshot(`
      <div
        className="flex banner mode"
      >
        <div
          className="message"
        >
          <pure(ErrorIcon)
            className="icon"
          />
          Some message
        </div>
        <div
          className="flex"
        >
          <a
            className="troubleShootingLink"
            href="https://www.kubeflow.org/docs/pipelines/troubleshooting"
          >
            Troubleshooting guide
          </a>
        </div>
      </div>
    `);
  });

  it('opens details dialog when button is clicked', () => {
    const tree = shallow(<Banner message='hello' additionalInfo='world' />);
    tree
      .find('WithStyles(Button)')
      .at(0)
      .simulate('click');
    expect(tree.state()).toHaveProperty('dialogOpen', true);
  });

  it('closes details dialog when cancel button is clicked', () => {
    const tree = shallow(<Banner message='hello' additionalInfo='world' />);
    tree
      .find('WithStyles(Button)')
      .at(0)
      .simulate('click');
    expect(tree.state()).toHaveProperty('dialogOpen', true);
    tree.find('#dismissDialogBtn').simulate('click');
    expect(tree.state()).toHaveProperty('dialogOpen', false);
  });

  it('calls refresh callback', () => {
    const spy = jest.fn();
    const tree = shallow(<Banner message='hello' refresh={spy} />);
    tree
      .find('.' + css.refreshButton)
      .at(0)
      .simulate('click');
    expect(spy).toHaveBeenCalled();
  });
});
