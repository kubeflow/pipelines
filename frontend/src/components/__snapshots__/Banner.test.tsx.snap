// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Banner defaults to error mode 1`] = `
<div
  className="flex banner mode"
>
  <div
    className="message"
  >
    <pure(ErrorIcon)
      className="icon"
    />
    Some message
  </div>
  <div
    className="flex"
  />
</div>
`;

exports[`Banner shows "Details" button and has dialog when there is additional info 1`] = `
<div
  className="flex banner mode"
>
  <div
    className="message"
  >
    <pure(ErrorIcon)
      className="icon"
    />
    Some message
  </div>
  <div
    className="flex"
  >
    <WithStyles(Button)
      className="button detailsButton"
      onClick={[Function]}
    >
      Details
    </WithStyles(Button)>
  </div>
  <WithStyles(Dialog)
    onClose={[Function]}
    open={false}
  >
    <WithStyles(DialogTitle)>
      An error occurred
    </WithStyles(DialogTitle)>
    <WithStyles(DialogContent)
      className="prewrap"
    >
      More info
    </WithStyles(DialogContent)>
    <WithStyles(DialogActions)>
      <WithStyles(Button)
        id="dismissDialogBtn"
        onClick={[Function]}
      >
        Dismiss
      </WithStyles(Button)>
    </WithStyles(DialogActions)>
  </WithStyles(Dialog)>
</div>
`;

exports[`Banner shows "Refresh" button when passed a refresh function 1`] = `
<div
  className="flex banner mode"
>
  <div
    className="message"
  >
    <pure(ErrorIcon)
      className="icon"
    />
    Some message
  </div>
  <div
    className="flex"
  >
    <WithStyles(Button)
      className="button refreshButton"
      onClick={[Function]}
    >
      Refresh
    </WithStyles(Button)>
  </div>
</div>
`;

exports[`Banner uses error mode when instructed 1`] = `
<div
  className="flex banner mode"
>
  <div
    className="message"
  >
    <pure(ErrorIcon)
      className="icon"
    />
    Some message
  </div>
  <div
    className="flex"
  />
</div>
`;

exports[`Banner uses warning mode when instructed 1`] = `
<div
  className="flex banner mode"
>
  <div
    className="message"
  >
    <pure(WarningIcon)
      className="icon"
    />
    Some message
  </div>
  <div
    className="flex"
  />
</div>
`;
